### DISCUSSION TITLE
Tracing a single websocket message

### DISCUSSION THREAD

**Discussion by danielsbird** [anycable/anycable#236] [Q&A] [answered]
Hi there,

I want to assign a unique identifier to each websocket message received by anycable-go from the browser and have anycable-go forward that unique identifier to the Anycable gRPC server. 

I read https://github.com/anycable/anycable-go/blob/master/docs/tracing.md and see anycable-go either assigns a random unique `sid` or uses the value provided in `X-Request-ID` HTTP header to each websocket connection and passes it with requests to RPC service. My understanding is that both the `sid` and the `X-Request-ID` HTTP header are not unique per websocket message. Instead, they are unique per websocket connection. So if the browser sends 2 messages over the same websocket connection then both of those messages will have the same `sid` or `X-Request-ID`.

I believe having a unique identifier for each websocket message will make it easier for me to trace requests that originate as websocket messages and flow from Anycable to the rest of my application's systems.

Is this possible now? If not, will you implement such a feature or accept a pull request?

**Comment by palkan** [comment, marked as answer]
Hey!

>  My understanding is that both the sid and the X-Request-ID HTTP header are not unique per websocket message. Instead, they are unique per websocket connection.

Correct.

> I believe having a unique identifier for each websocket message will make it easier for me to trace requests that originate as websocket messages and flow from Anycable to the rest of my application's systems.

> Is this possible now? If not, will you implement such a feature or accept a pull request?

Currently, it's not possible. The primary reason is the the underlying protocol we use, Action Cable, has no such capabilities. We can extend it though (like we already did for some features).

The way I see it as follows:

1. The client assigns a unique ID to each outgoing command (or the user provides one when invoking the command). For that, we must extend the current command format and add some `trace_id` or whatever.

2. AnyCable server retrieves the `trace_id` from the client message and send it over the RPC via the RPC request metadata (similarly to `Sid`).

3. The application may access the `trace_id` from the request metadata (we can store it in the `request.env["anycable.trace_id"]` in Ruby/Rails).

If there is no need in the client-originated trace ID, you can attach IDs at the RPC level (e.g., via an AnyCable middleware) to distinguish commands at the application level.

**Reply by danielsbird** [reply]
Thanks for your answer!
